README.md
README file for code associated with:

A bias-variance trade-off governs individual differences in on-line learning in an unpredictable environment

Christopher M. Glaze, Alexandre L.S. Filipowicz, Joseph W. Kable, Vijay Balasubramanian, and Joshua I. Gold

Nature Human Behavior
2018

Model-fitting routines. These routines work but are not organized or commented as clearly as the figure-generating code. Please contact the authors with questions.

Notes (from C.Glaze):

Routines for fitting data and saved parameter estimates.

Creates *.mat files beginning with ‘allparam’ that contain fit parameters, and have names with basic syntax
‘allparam_fixedvarfits_[learning model]_[fixed parameters].mat’. (‘fixedvarfits’ refers to the fact that subjective generative variance is written as a constant linear function of the task generative variance vs. previous learning models CG analyzed in which variance is learned online within each session).

*.m files beginning with ‘analyze’ contain routines for fitting parameters with syntax ‘analyze_everyone_2AFC_[learning model].mat’

All files ending in ‘crossval’ are dedicated to estimating particular learning model parameters in one of two randomly generated partitions of experimental sessions for each subject. Session assignments to partitions were random but fixed for all analysis so that results could be most accurately compared across models. Partition information is in ‘random_session_assignments.mat’.

All fits included two basic kinds of structures:
fixedparamstrct: 
	-params0: initial parameter estimates
	-lb, ub: lower, upper boundaries on parameter estimates
	-free: binary vector indicating which parameters are free. those that are not are fixed at the 	params0 values.

paramstrct: structure with fixedparamstrct information and additional fields:
	-params: fit parameters
	-LL: model LL 
	-LLrl (sampling model fits only): model LL re-calculated after fitting was complete

NOTE: I’ve only included final models for manuscript, in which the task generative variance is assumed to be known and used correctly. This is based on (1) the fact that subjects were shown the true Gaussian distributions before each session and directly instructed on each variance, (2) allowed 120 practice trials before each experimental session and (3) previous model analysis showing that allowing subjective generative variance to freely vary resulted in lower model evidence.

MODELS:

***cmex files***
-All c-mex files contain the basic learning algorithms investigated for the research, pre-compiled for Mac with the default gcc compiler in the Xcode associated with El Captain. Windows users will have to re-compile: 

-Bayesian_mixedXX.c has the ideal observer algorithm, with each version storing different kinds of information (e.g. Bayesian_6.c only stores and returns posterior over state space for speed of calculation in fitting algorithm).

-fixedLR_learnHXX.c has a reduced learning algorithm that makes point-estimates of hazard rate trial by trial with a delta rule that operates on change point probability (2nd moment of posterior in state space).

-particle_filter_learnHXX.c: novel sampling method for learning hazard rates as reported in manuscript.

-dsprt_mixed_Hvc2.c: non-learning algorithm implementing HMM solution with a hazard rate that can vary trial-by-trial, used for fitting the initial regression model and blocked estimates of subjective hazard rate.

***geffit_XXX.m***
m-files that get the data likelihood for a given model. Each of these take in arrays of task and choice data so that entire subjects can be fit all at once. Each uses Matlab’s ‘parfor’ command to loop through arrays for speed. Those without the parallel processing toolbox can simply change this to ‘for’.

***fit_XXX.m***
m-files that do the actual fits of choice data to learning models. Every function has the same syntax:
-first 3 arguments are arrays of task and choice data, 
-2nd two indicate task generative variances and the average data mean (always 30 in this task)
-last argument is a structure fixedparamstrct with the following parallel fields:
	-params0: initial parameter estimates
	-lb, ub: lower, upper boundaries on parameter estimates
	-free: binary vector indicating which parameters are free. those that are not are fixed at the 	params0 values.


RUNNING THE SCRIPTS

Below are the instructions on how to run the model fitting for the models used in Glaze et al. 2018 


1) All of the model scripts are found in ../Models/ Go to this folder and modify line 16 of "getDataInfo.m" by putting in the correct directory for your computer

2) All of the data files for this paper are contained in the Data/ folder. See README in that section to get breakdown of the data structures

3) The models and fitting routines are found in the Models/ folder. The naming conventions are explained the that section's README.md file.

4) To fit a model to data, run the appropriate "analyze_everyone_2AFC_XX" file, where XX identifies the model you want to fit. As long as the "getDataInfo.m" has the path corrected, this should run.

5) Each "analyze" file returns a "paramstruct_" structure with all of the parameter. It is organized as follows:
	- varnms: parameter names
	- params0: initial parameters for the fitting
	- ub: parameter upper bounds
	- lb: parameter lower bounds
	- free: which parameters should be considered free(1) and which considered fixed (0)
	- params: parameters for each subject
	- LL: log likelihood
	


changes.txt
## This version of the task was created way before the MTurk version so there are numerous things that differ.

(1) This is the prediction version so subjects are asked to choose which chicken will lay the next egg.

(2) The chickens are further apart so there is less overlap. 
	Left Chicken Position is (x=-342, y=0), 
	Right Chicken Position is (x=342, y=0)
	Block definitions are as follows:
	Block 1 -> Hrate=0.05, Sigma=150
	Block 2 -> Hate=0.05, Sigma=200
	Block 3 -> Hrate=0.2, Sigma=150
	Block 4 -> Hrate=0.2, Sigma=200
	Block 5 -> Hrate=0.95, Sigma=150
	Block 6 -> Hrate=0.95, Sigma=200

(3) Feedback is given every trial (there is a bar at the top showing their points progress but not the actual points)

(4) Each subject gets either of the 3 patterns randomly. That pattern is the same for T1 and T2.
The different patterns are just the different ordering of the blocks.

Pattern 1: (http://brainworkout.paulus.libr.net/?task=chicken&pattern=1) ->  table134.txt
Pattern 2: (http://brainworkout.paulus.libr.net/?task=chicken&pattern=2) -> table145.txt
Pattern 3: (http://brainworkout.paulus.libr.net/?task=chicken&pattern=3) -> table4.txt

createFigures.R
library(ggplot2)
library(randomcoloR)
library(psych)

# Set working directory
setwd("/Volumes/T1000/Analysis/kforthman/Chicken Task/Chicken_code")

# Read in the data.
data <- read.csv("Data/Analysis/adaptivityModelFits.csv")
# Change percent to decimal
data[,'pct'] = data[,'pct']/100

# Now load in the pattern information
pat_data <- read.csv("Data/blockPatterns.csv")
pat_T1 <- pat_data[which(pat_data[,'t']=='T1'),]

# These are the h and sig for each block, in block order.
sig <- c(33,140,33,140,33,140)
h <- c(0.05,0.05,0.2,0.2,0.95,0.95)
# Below are some clearer definitions of what each h and sig entails
sigdef <- c("Small Egg Distribution", "Large Egg Distribution","Small Egg Distribution", "Large Egg Distribution","Small Egg Distribution", "Large Egg Distribution")
hdef <- c("Very Infrequent Switching","Very Infrequent Switching", "Infrequent Switching", "Infrequent Switching", "Very Frequent Switching", "Very Frequent Switching")

# how the pattern will be represented in the plot
pattern <- c("turquoise3", "seagreen4", "slateblue", "maroon")


# Block-wise subjective (fit) H vs objective H. Solid line is a least-squares fit. 
png(paste0("Figures/H_Default_Comparison_to_H_true.png"))
plot(data$H_true, data$H_subjective, xlim = c(0,1), ylim = c(0,1))
tru2def_fit <- lm(data$H_subjective~data$H_true)
abline(tru2def_fit)
dev.off()

t_cat <- levels(unique(data$trial))
sig_cat <- unique(data$sigma)
for(i in 1:length(t_cat)){
  for(j in 1:length(sig_cat)){
    index <- which(data$trial == t_cat[i] & data$sigma == sig_cat[j])
    png(paste0("Figures/H_Default_Comparison_to_H_true-", t_cat[i], "-", sig_cat[j],".png"))
    plot(jitter(data$H_true[index]), data$H_subjective[index], xlim = c(0,1), ylim = c(0,1), 
         main = paste0("Trial: ", t_cat[i], ", Sigma: ", sig_cat[j]),
         xlab = "True H",
         ylab = "Subjective H")
    tru2def_fit <- lm(data$H_subjective[index]~data$H_true[index])
    abline(tru2def_fit)
    #text(.9,0,paste0("Adj R2 = ", round(signif(summary(tru2def_fit)$adj.r.squared, 5), 3)))
    dev.off()
  }
}

# Block-wise subjective (fit) H vs objective H.
id <- levels(unique(data$ID))
png(paste0("Figures/H_Default_Comparison_to_H_true-spaghetti.png"))
plot(data$H_true, data$H_subjective, xlim = c(0,1), ylim = c(0,1))
for(i in 1:length(id)){
  index <- which(data$ID == id[i])
  tru2def_fit <- lm(data$H_subjective[index]~data$H_true[index])
  abline(tru2def_fit, col = randomColor(1))
}
dev.off()

# The following plots show the default H compared at T1 and T2. The wheels marking each point indicate the percent correct. The length of the + spokes indicate percent correct at T1 and the length of the x spokes indicate percent correct at T2. If the spokes reach the edge of the wheel, that means the participant had ~100% accuracy. The pink line is the least-squares fit. The dashed line is just a 45 degree line.
for(i in 1:6){
  
  t1 <- which(data[,2] == 'T1' & data[,3]==i)
  t2 <- which(data[,2] == 'T2' & data[,3]==i)
  
  fit <- lm(data[t2,6]~data[t1,6])
  
  icc <- ICC(cbind(data[t1,6],data[t2,6]))
  
  gg <- ggplot(fit$model, aes_string(x = names(fit$model)[2], y = names(fit$model)[1])) + 
    stat_smooth(method = "lm", se = F, colour = "pink") + # Linear regression
    geom_abline(intercept = 0, colour = "slateblue4", linetype = "dashed") + # 45 degree line
    geom_point(colour = pattern[pat_T1[,'pattern']], size = data[t1,'pct']*4, pch = 3) + # wheels
    geom_point(colour = pattern[pat_T1[,'pattern']], size = data[t2,'pct']*4, pch = 4) +
    geom_point(colour = pattern[pat_T1[,'pattern']], size = 6, pch = 1)+ scale_fill_continuous(guide = guide_legend()) +
    # geom_point(colour = rgb(data[t1,11], 0, data[t2,11]), pch = 16, size = 5) + #color
    labs(title = paste0("Block ", i, ": H = ", h[i], ", sigma = ", sig[i]),
         caption = paste0("Adj R2 = ", round(signif(summary(fit)$adj.r.squared, 5), 3),
                          ", Intercept =", round(signif(fit$coef[[1]],5 ), 3),
                          ", Slope =", round(signif(fit$coef[[2]], 5), 3),
                          ", P =", round(signif(summary(fit)$coef[2,4], 5), 3),
                          ",\nICC (single absolute) = ", round(icc$results$ICC[1],3),
                          ", ICC (single fixed) = ", round(icc$results$ICC[3], 3)),
         subtitle = paste0(sigdef[i], ", ", hdef[i]))+
    xlim(0,1)+
    ylim(0,1)+
    xlab("T1 Subjective H")+
    ylab("T2 Subjective H")  + theme(legend.position = "bottom")
  
  png(paste0("Figures/H_Default_Comparison-Block_", i, ".png"),
      bg = "transparent")
  plot(gg)
  dev.off()
  
  
  #cool color palette graph
  # x <- rep(seq(0,1,length.out = 100),100)
  # y <- seq(0,1,length.out = 10000)
  # plot(x,y)
  # dummy <- data.frame(x,y)
  # 
  # gg <- ggplot(dummy, aes_string(x = x, y = y)) +
  #   geom_point(colour = rgb(x, 0, y), pch = 16, size = 5) + #color
  #   xlim(0,1)+
  #   ylim(0,1)+
  #   xlab("T1 Accuracy")+
  #   ylab("T2 Accuracy")
  # 
  # plot(gg)
  
  #plot just pct
  # fit <- lm(data[t2,11]~data[t1,11])
  # 
  # gg <- ggplot(fit$model, aes_string(x = names(fit$model)[2], y = names(fit$model)[1])) + 
  #   geom_point(colour = "turquoise4") +
  #   stat_smooth(method = "lm", se = F, colour = "turquoise") +
  #   labs(title = paste0("Block ", i, ": H = ", h[i], ", sigma = ", sig[i]),
  #        caption = paste0("Adj R2 = ",signif(summary(fit)$adj.r.squared, 5),
  #                         ", Intercept =",signif(fit$coef[[1]],5 ),
  #                         ", Slope =",signif(fit$coef[[2]], 5),
  #                         ", P =",signif(summary(fit)$coef[2,4], 5)),
  #        subtitle = paste0(sigdef[i], ", ", hdef[i]))+
  #   xlim(0,100)+
  #   ylim(0,100)+
  #   xlab("T1 Percent Correct")+
  #   ylab("T2 Percent Correct")
  # 
  # plot(gg)
}

# Distribution of parameter estimates across subjects
data <- read.csv("Data/Analysis/adaptivityModelFits.csv")
hist(as.matrix(data["noise_in_DV"]), breaks=30)
hist(as.matrix(data["lapse_rate"]), breaks=30)
hist(as.matrix(data["H_subjective"]))


